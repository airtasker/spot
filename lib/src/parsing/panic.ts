import ts from "typescript";

export function panic(nodeOrMessage: ts.Node | string) {
  if (typeof nodeOrMessage === "string") {
    return new Error(nodeOrMessage);
  }
  let syntaxKind = "unknown";
  for (const [key, value] of Object.entries(ts.SyntaxKind)) {
    if (nodeOrMessage.kind === value) {
      syntaxKind = key;
      break;
    }
  }
  return new Error(syntaxKind + ": " + JSON.stringify(nodeOrMessage, null, 2));
}
