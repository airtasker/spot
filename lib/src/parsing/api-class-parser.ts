import * as ts from "typescript";
import { isHttpContentType, isHttpMethod } from "../lib";
import {
  Api,
  DynamicPathComponent,
  Endpoint,
  Headers,
  PathComponent,
  QueryParamComponent,
  SpecificError,
  Type,
  VOID
} from "../models";
import {
  extractMultipleDecorators,
  extractSingleDecorator
} from "./decorators";
import {
  extractLiteral,
  isNumericLiteral,
  isObjectLiteral,
  isStringLiteral,
  Literal
} from "./literal-parser";
import { panic } from "./panic";
import { extractType } from "./type-parser";

/**
 * Parses a top-level API class definition and the endpoints it defines, such as:
 * ```
 * @api()
 * class Api {
 *   @endpoint({
 *     method: "POST",
 *     path: "/users"
 *   })
 *   createUser(@request req: CreateUserRequest): CreateUserResponse {
 *     throw "contract";
 *   }
 * }
 * ```
 */
export function parseApiClass(
  sourceFile: ts.SourceFile,
  classDeclaration: ts.ClassDeclaration,
  api: Api
): void {
  for (const member of classDeclaration.members) {
    if (ts.isMethodDeclaration(member)) {
      parseEndpointMethod(sourceFile, member, api);
    }
  }
}

/**
 * Parses a method of an API class definition, such as:
 * ```
 * @endpoint({
 *   method: "POST",
 *   path: "/users"
 * })
 * createUser(@request req: CreateUserRequest): CreateUserResponse {
 *   throw "contract";
 * }
 * ```
 *
 * Methods that do not have an @endpoint() decorator will be ignored.
 *
 * @param sourceFile The TypeScript source file.
 * @param methodDeclaration A method declaration.
 */
function parseEndpointMethod(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  api: Api
): void {
  const endpointDecorator = extractSingleDecorator(
    sourceFile,
    methodDeclaration,
    "endpoint"
  );
  if (!endpointDecorator) {
    return;
  }
  const endpointName = methodDeclaration.name.getText(sourceFile);
  if (api.endpoints[endpointName]) {
    throw panic(
      `Found multiple definitions of the same endpoint ${endpointName}`
    );
  }
  const genericErrorDecorator = extractSingleDecorator(
    sourceFile,
    methodDeclaration,
    "genericError"
  );
  const specificErrorDecorators = extractMultipleDecorators(
    sourceFile,
    methodDeclaration,
    "specificError"
  );
  if (endpointDecorator.arguments.length !== 1) {
    throw panic(
      `Expected exactly one argument for @endpoint(), got ${
        endpointDecorator.arguments.length
      }`
    );
  }
  const endpointDescriptionExpression = endpointDecorator.arguments[0];
  const endpointDescription = extractLiteral(
    sourceFile,
    endpointDescriptionExpression
  );
  if (!isObjectLiteral(endpointDescription)) {
    throw panic(
      `@endpoint() expects an object literal, got this instead: ${endpointDescriptionExpression.getText(
        sourceFile
      )}`
    );
  }
  const methodLiteral = endpointDescription.properties["method"];
  const pathLiteral = endpointDescription.properties["path"];
  const requestContentTypeLiteral =
    endpointDescription.properties["requestContentType"];
  const successStatusCodeLiteral =
    endpointDescription.properties["successStatusCode"];

  if (!isStringLiteral(methodLiteral)) {
    throw panic(
      `Invalid method in endpoint description: ${endpointDescriptionExpression.getText(
        sourceFile
      )}`
    );
  }
  const method = methodLiteral.text;
  if (!isHttpMethod(method)) {
    throw panic(`${method} is not a valid HTTP method`);
  }
  if (!isStringLiteral(pathLiteral)) {
    throw panic(
      `Invalid path in endpoint description: ${endpointDescriptionExpression.getText(
        sourceFile
      )}`
    );
  }

  let requestContentType = "application/json";
  if (requestContentTypeLiteral) {
    if (!isStringLiteral(requestContentTypeLiteral)) {
      throw panic(
        `Invalid request content type in endpoint description: ${endpointDescriptionExpression.getText(
          sourceFile
        )}`
      );
    }
    requestContentType = requestContentTypeLiteral.text;
  }
  if (!isHttpContentType(requestContentType)) {
    throw panic(`${method} is not a valid HTTP content type`);
  }

  let successStatusCode;
  if (successStatusCodeLiteral) {
    if (!isNumericLiteral(successStatusCodeLiteral)) {
      throw panic(
        `Invalid success status code in endpoint description: ${endpointDescriptionExpression.getText(
          sourceFile
        )}`
      );
    }
    successStatusCode = parseInt(successStatusCodeLiteral.text);
  }
  const path = pathLiteral.text;
  const pathComponents: PathComponent[] = [];
  const dynamicPathComponents: { [name: string]: DynamicPathComponent } = {};
  if (path.length > 0) {
    let componentStartPosition = 0;
    do {
      if (path.charAt(componentStartPosition) === ":") {
        // The parameter name extends until a character that isn't a valid name.
        const nextNonNamePositionRelative = path
          .substr(componentStartPosition + 1)
          .search(/[^a-z0-9_]/gi);
        const dynamicPathComponent: DynamicPathComponent = {
          kind: "dynamic",
          name: path.substr(
            componentStartPosition + 1,
            nextNonNamePositionRelative === -1
              ? undefined
              : nextNonNamePositionRelative
          ),
          type: VOID,
          description: ""
        };
        pathComponents.push(dynamicPathComponent);
        dynamicPathComponents[dynamicPathComponent.name] = dynamicPathComponent;
        componentStartPosition =
          nextNonNamePositionRelative === -1
            ? -1
            : componentStartPosition + 1 + nextNonNamePositionRelative;
      } else {
        // The static component extends until the next parameter, which starts with ":".
        const nextColumnPosition = path.indexOf(":", componentStartPosition);
        pathComponents.push({
          kind: "static",
          content: path.substring(
            componentStartPosition,
            nextColumnPosition === -1 ? undefined : nextColumnPosition
          )
        });
        componentStartPosition = nextColumnPosition;
      }
    } while (componentStartPosition !== -1);
  }
  let requestType: Type = VOID;
  const headers: Headers = {};
  const queryParams: QueryParamComponent[] = [];
  const queryParamComponents: { [name: string]: QueryParamComponent } = {};
  for (const parameter of methodDeclaration.parameters) {
    const requestDecorator = extractSingleDecorator(
      sourceFile,
      parameter,
      "request"
    );
    const pathParamDecorator = extractSingleDecorator(
      sourceFile,
      parameter,
      "pathParam"
    );
    const headerDecorator = extractSingleDecorator(
      sourceFile,
      parameter,
      "header"
    );
    const queryParamDecorator = extractSingleDecorator(
      sourceFile,
      parameter,
      "queryParam"
    );
    if (parameter.questionToken) {
      throw panic(
        `Question tokens are not allowed in parameter definitions. Please use Optional<...> to be specific. Offending parameter: ${parameter.getText(
          sourceFile
        )}`
      );
    }
    if (!ts.isIdentifier(parameter.name)) {
      throw panic(
        `Expected a plain identifier for endpoint parameter name, got this instead: ${parameter.getText(
          sourceFile
        )}`
      );
    }
    if (!parameter.type) {
      throw panic(
        `Expected a type for endpoint parameter: ${parameter.getText(
          sourceFile
        )}`
      );
    }
    const type = extractType(sourceFile, parameter.type);
    if (requestDecorator) {
      requestType = type;
    } else if (pathParamDecorator) {
      const name = parameter.name.getText(sourceFile);
      if (dynamicPathComponents[name]) {
        dynamicPathComponents[name].type = type;
      } else {
        throw panic(
          `Found a path parameter that isn't present in path. Expected one of [${Object.keys(
            dynamicPathComponents
          ).join(", ")}], got this instead: ${name}`
        );
      }
      if (pathParamDecorator.arguments.length > 1) {
        throw panic(
          `Expected exactly one or no arguments for @pathParam(), got ${
            pathParamDecorator.arguments.length
          }`
        );
      }
      if (pathParamDecorator.arguments.length === 1) {
        const pathParamDescription = extractLiteral(
          sourceFile,
          pathParamDecorator.arguments[0]
        );
        if (!isObjectLiteral(pathParamDescription)) {
          throw panic(
            `@pathParam() expects an object literal, got this instead: ${pathParamDecorator.arguments[0].getText(
              sourceFile
            )}`
          );
        }
        const descriptionProperty =
          pathParamDescription.properties["description"];
        if (!descriptionProperty || !isStringLiteral(descriptionProperty)) {
          throw panic(
            `@pathParam() expects a string description, got this instead: ${pathParamDecorator.arguments[0].getText(
              sourceFile
            )}`
          );
        }
        dynamicPathComponents[name].description = descriptionProperty.text;
      }
    } else if (queryParamDecorator) {
      const name = parameter.name.getText(sourceFile);

      if (queryParamComponents[name]) {
        throw panic(`Found multiple query parameters named ${name}`);
      } else {
        const queryParamComponent: QueryParamComponent = {
          name: name,
          description: "",
          type: type
        };
        queryParams.push(queryParamComponent);
        queryParamComponents[queryParamComponent.name] = queryParamComponent;
      }

      if (queryParamDecorator.arguments.length > 1) {
        throw panic(
          `Expected exactly one or no arguments for @queryParam(), got ${
            queryParamDecorator.arguments.length
          }`
        );
      }
      if (queryParamDecorator.arguments.length === 1) {
        const queryParamDescription = extractLiteral(
          sourceFile,
          queryParamDecorator.arguments[0]
        );
        if (!isObjectLiteral(queryParamDescription)) {
          throw panic(
            `@queryParam() expects an object literal, got this instead: ${queryParamDecorator.arguments[0].getText(
              sourceFile
            )}`
          );
        }
        const descriptionProperty =
          queryParamDescription.properties["description"];
        if (!descriptionProperty || !isStringLiteral(descriptionProperty)) {
          throw panic(
            `@queryParam() expects a string description, got this instead: ${queryParamDecorator.arguments[0].getText(
              sourceFile
            )}`
          );
        }
        queryParamComponents[name].description = descriptionProperty.text;
      }
    } else if (headerDecorator) {
      const name = parameter.name.getText(sourceFile);
      if (headers[name]) {
        throw panic(`Found multiple headers named ${name}`);
      }
      if (headerDecorator.arguments.length !== 1) {
        throw panic(
          `Expected exactly one argument for @header(), got ${
            headerDecorator.arguments.length
          }`
        );
      }
      const headerDescription = extractLiteral(
        sourceFile,
        headerDecorator.arguments[0]
      );
      if (!isObjectLiteral(headerDescription)) {
        throw panic(
          `@header() expects an object literal, got this instead: ${headerDecorator.arguments[0].getText(
            sourceFile
          )}`
        );
      }
      const nameProperty = headerDescription.properties["name"];
      if (!nameProperty || !isStringLiteral(nameProperty)) {
        throw panic(
          `@header() expects a string name, got this instead: ${headerDecorator.arguments[0].getText(
            sourceFile
          )}`
        );
      }
      const descriptionProperty = headerDescription.properties["description"];
      let description = "";
      if (descriptionProperty) {
        if (!isStringLiteral(descriptionProperty)) {
          throw panic(
            `@header() expects a string description, got this instead: ${headerDecorator.arguments[0].getText(
              sourceFile
            )}`
          );
        }
        description = descriptionProperty.text;
      }
      headers[name] = {
        headerFieldName: nameProperty.text,
        description,
        type
      };
    } else {
      throw panic(
        `Found a parameter without @request, @pathParam or @header: ${parameter.getText(
          sourceFile
        )}`
      );
    }
  }
  let responseType: Type = VOID;
  if (methodDeclaration.type) {
    if (
      ts.isTypeReferenceNode(methodDeclaration.type) &&
      ts.isIdentifier(methodDeclaration.type.typeName) &&
      methodDeclaration.type.typeName.escapedText === "Promise"
    ) {
      if (
        !methodDeclaration.type.typeArguments ||
        methodDeclaration.type.typeArguments.length !== 1
      ) {
        throw panic(
          `Expected Promise<...>, got this instead: ${methodDeclaration.type.getText(
            sourceFile
          )}`
        );
      }
      const promisedType = methodDeclaration.type.typeArguments[0];
      responseType = extractType(sourceFile, promisedType);
    } else {
      responseType = extractType(sourceFile, methodDeclaration.type);
    }
  }
  let genericErrorType: Type = VOID;
  let specificErrorTypes: {
    [name: string]: SpecificError;
  } = {};
  if (genericErrorDecorator) {
    if (genericErrorDecorator.typeParameters.length !== 1) {
      throw panic(
        `Expected exactly one type parameter for @genericError(), got ${
          genericErrorDecorator.typeParameters.length
        }`
      );
    }
    genericErrorType = extractType(
      sourceFile,
      genericErrorDecorator.typeParameters[0]
    );
  }
  for (const specificErrorDecorator of specificErrorDecorators) {
    if (specificErrorDecorator.typeParameters.length !== 1) {
      throw panic(
        `Expected exactly one type parameter for @specificError(), got ${
          specificErrorDecorator.typeParameters.length
        }`
      );
    }
    const errorResponseType = extractType(
      sourceFile,
      specificErrorDecorator.typeParameters[0]
    );
    if (specificErrorDecorator.arguments.length !== 1) {
      throw panic(
        `Expected exactly one argument for @specificError(), got ${
          specificErrorDecorator.arguments.length
        }`
      );
    }
    let errorDescription: Literal;
    if (specificErrorDecorator.arguments.length === 1) {
      errorDescription = extractLiteral(
        sourceFile,
        specificErrorDecorator.arguments[0]
      );
      if (!isObjectLiteral(errorDescription)) {
        throw panic(
          `@specificError() expects an object literal, got this instead: ${specificErrorDecorator.arguments[0].getText(
            sourceFile
          )}`
        );
      }
    } else {
      errorDescription = {
        kind: "object",
        properties: {}
      };
    }
    const name = errorDescription.properties["name"];
    if (!name || !isStringLiteral(name)) {
      throw panic(
        `@specificError() expects a string name, got this instead: ${specificErrorDecorator.arguments[0].getText(
          sourceFile
        )}`
      );
    }
    const statusCode = errorDescription.properties["statusCode"];
    if (!statusCode || !isNumericLiteral(statusCode)) {
      throw panic(
        `@specificError() expects a numeric status code, got this instead: ${specificErrorDecorator.arguments[0].getText(
          sourceFile
        )}`
      );
    }
    specificErrorTypes[name.text] = {
      // TODO: Ensure that the status is an integer.
      statusCode: parseInt(statusCode.text),
      type: errorResponseType
    };
  }
  const endpoint: Endpoint = {
    method,
    path: pathComponents,
    requestContentType,
    headers,
    queryParams,
    requestType,
    responseType,
    genericErrorType,
    specificErrorTypes,
    ...(successStatusCode ? { successStatusCode } : {})
  };
  api.endpoints[endpointName] = endpoint;
}
