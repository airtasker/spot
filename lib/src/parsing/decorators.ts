import * as ts from "typescript";
import { Type } from "../models";
import { extractLiteral, Literal } from "./literal-parser";
import { panic } from "./panic";
import { extractType } from "./type-parser";

/**
 * Extracts the arguments from a single decorator associated with a TypeScript node.
 *
 * For example with the following code:
 * ```
 * @myDecorator(1, 2, 3)
 * function abc() {}
 * ```
 *
 * Then calling extractSingleDecorator() on the function declaration node will return:
 * {
 *   typeParameters: [],
 *   arguments: [1, 2, 3]
 * }
 *
 * This will fail if the same decorator is attached to the same node multiple times.
 * Use extractMultipleDecorators() for that purpose instead.
 *
 * @returns null when there is no decorator with the given name attached to the node.
 */
export function extractSingleDecorator(
  sourceFile: ts.SourceFile,
  node: ts.Node,
  decoratorName: string
): Decorator | null {
  const decorators = extractMultipleDecorators(sourceFile, node, decoratorName);
  if (decorators.length === 1) {
    return decorators[0];
  } else if (decorators.length > 1) {
    throw panic(
      `Expected a single @${decoratorName} decorator, found ${
        decorators.length
      }`
    );
  } else {
    return null;
  }
}

/**
 * Extracts the arguments from a repeated decorator associated with a TypeScript node.
 *
 * For example with the following code:
 * ```
 * @myDecorator(1, 2, 3)
 * @myDecorator("a", "b", "c")
 * function abc() {}
 * ```
 *
 * Then calling extractMultipleDecorators() on the function declaration node will return:
 * [
 *   {
 *     typeParameters: [],
 *     arguments: [1, 2, 3]
 *   },
 *   {
 *     typeParameters: [],
 *     arguments: ["a", "b", "c"]
 *   }
 * ]
 */
export function extractMultipleDecorators(
  sourceFile: ts.SourceFile,
  node: ts.Node,
  decoratorName: string
): Decorator[] {
  const decorators: Decorator[] = [];
  if (node.decorators) {
    for (const decorator of node.decorators) {
      if (
        ts.isIdentifier(decorator.expression) &&
        decorator.expression.getText(sourceFile) === decoratorName
      ) {
        decorators.push({
          typeParameters: [],
          arguments: []
        });
      } else if (ts.isCallExpression(decorator.expression)) {
        if (
          ts.isIdentifier(decorator.expression.expression) &&
          decorator.expression.expression.getText(sourceFile) === decoratorName
        ) {
          decorators.push({
            typeParameters: [...(decorator.expression.typeArguments || [])].map(
              t => extractType(sourceFile, t)
            ),
            arguments: [...decorator.expression.arguments].map(e =>
              extractLiteral(sourceFile, e)
            )
          });
        }
      }
    }
  }
  return decorators;
}

export interface Decorator {
  typeParameters: Type[];
  arguments: Literal[];
}
