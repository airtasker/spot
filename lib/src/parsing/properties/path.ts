import ts from "typescript";
import { DynamicPathComponent, PathComponent, VOID } from "../../models";
import { extractSingleDecorator } from "../decorators";
import {
  isObjectLiteral,
  isStringLiteral,
  ObjectLiteral
} from "../literal-parser";
import { panic } from "../panic";
import { extractParameterType } from "./parameter-type";

/**
 * Returns the path of an endpoint.
 *
 * For example:
 * ```
 * @endpoint({
 *   method: "GET",
 *   path: "/users/:userId"
 * })
 * getUser(@pathParam() userId: string): User {
 *   ...
 * }
 * ```
 *
 * will return a list of path components with:
 * - a static component ("/users/")
 * - a dynamic component named userId of type string
 */
export function extractPath(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  endpointDescription: ObjectLiteral
): PathComponent[] {
  const pathLiteral = endpointDescription.properties["path"];
  if (!isStringLiteral(pathLiteral)) {
    throw panic(
      `Invalid path in endpoint description: ${methodDeclaration.getText(
        sourceFile
      )}`
    );
  }
  const path = pathLiteral.text;
  const pathComponents: PathComponent[] = [];
  const dynamicPathComponents: { [name: string]: DynamicPathComponent } = {};
  if (path.length > 0) {
    let componentStartPosition = 0;
    do {
      if (path.charAt(componentStartPosition) === ":") {
        // The parameter name extends until a character that isn't a valid name.
        const nextNonNamePositionRelative = path
          .substr(componentStartPosition + 1)
          .search(/[^a-z0-9_]/gi);
        const dynamicPathComponent: DynamicPathComponent = {
          kind: "dynamic",
          name: path.substr(
            componentStartPosition + 1,
            nextNonNamePositionRelative === -1
              ? undefined
              : nextNonNamePositionRelative
          ),
          type: VOID,
          description: ""
        };
        pathComponents.push(dynamicPathComponent);
        dynamicPathComponents[dynamicPathComponent.name] = dynamicPathComponent;
        componentStartPosition =
          nextNonNamePositionRelative === -1
            ? -1
            : componentStartPosition + 1 + nextNonNamePositionRelative;
      } else {
        // The static component extends until the next parameter, which starts with ":".
        const nextColumnPosition = path.indexOf(":", componentStartPosition);
        pathComponents.push({
          kind: "static",
          content: path.substring(
            componentStartPosition,
            nextColumnPosition === -1 ? undefined : nextColumnPosition
          )
        });
        componentStartPosition = nextColumnPosition;
      }
    } while (componentStartPosition !== -1);
  }
  for (const parameter of methodDeclaration.parameters) {
    processPathParameter(
      sourceFile,
      methodDeclaration,
      parameter,
      dynamicPathComponents
    );
  }
  return pathComponents;
}

function processPathParameter(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  parameter: ts.ParameterDeclaration,
  dynamicPathComponents: { [name: string]: DynamicPathComponent }
) {
  const pathParamDecorator = extractSingleDecorator(
    sourceFile,
    parameter,
    "pathParam"
  );
  if (!pathParamDecorator) {
    return;
  }
  const name = parameter.name.getText(sourceFile);
  if (dynamicPathComponents[name]) {
    dynamicPathComponents[name].type = extractParameterType(
      sourceFile,
      parameter
    );
  } else {
    throw panic(
      `Found a path parameter that isn't present in path. Expected one of [${Object.keys(
        dynamicPathComponents
      ).join(", ")}], got this instead: ${name}`
    );
  }
  if (pathParamDecorator.arguments.length > 1) {
    throw panic(
      `Expected exactly one or no arguments for @pathParam(), got ${
        pathParamDecorator.arguments.length
      }`
    );
  }
  if (pathParamDecorator.arguments.length === 1) {
    const pathParamDescription = pathParamDecorator.arguments[0];
    if (!isObjectLiteral(pathParamDescription)) {
      throw panic(
        `@pathParam() expects an object literal, got this instead: ${methodDeclaration.getText(
          sourceFile
        )}`
      );
    }
    const descriptionProperty = pathParamDescription.properties["description"];
    if (!descriptionProperty || !isStringLiteral(descriptionProperty)) {
      throw panic(
        `@pathParam() expects a string description, got this instead: ${methodDeclaration.getText(
          sourceFile
        )}`
      );
    }
    dynamicPathComponents[name].description = descriptionProperty.text;
  }
}
