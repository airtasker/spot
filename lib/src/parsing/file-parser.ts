import * as fs from "fs-extra";
import * as path from "path";
import * as ts from "typescript";
import { Api } from "../models";
import { validate } from "../validator";
import { parseApiClass } from "./api-class-parser";
import { extractSingleDecorator } from "./decorators";
import { panic } from "./panic";
import { extractObjectType, extractType } from "./type-parser";

/**
 * Parses a TypeScript source file, as well as any other TypeScript files it imports recursively.
 *
 * This should be smart enough not to get stuck into an infinite loop for circular dependencies.
 */
export async function parsePath(sourcePath: string): Promise<Api> {
  const api: Api = {
    endpoints: {},
    types: {}
  };
  await parseFileRecursively(api, new Set(), sourcePath);
  const errors = validate(api);
  if (errors.length > 0) {
    throw panic(errors.join("\n"));
  }
  return api;
}

async function parseFileRecursively(
  api: Api,
  visitedPaths: Set<string>,
  sourcePath: string
): Promise<void> {
  if (!(await fs.existsSync(sourcePath))) {
    if (await fs.existsSync(sourcePath + ".ts")) {
      sourcePath += ".ts";
    } else {
      throw panic(`No source file found at ${sourcePath}`);
    }
  }
  if (
    visitedPaths.has(sourcePath) ||
    path.resolve(sourcePath).startsWith(__dirname)
  ) {
    return;
  } else {
    visitedPaths.add(sourcePath);
  }
  const fileContent = await fs.readFile(sourcePath, "utf8");
  const sourceFile = ts.createSourceFile(
    path.basename(sourcePath),
    fileContent,
    ts.ScriptTarget.Latest
  );
  for (const statement of sourceFile.statements) {
    if (ts.isImportDeclaration(statement)) {
      if (!ts.isStringLiteral(statement.moduleSpecifier)) {
        throw panic(
          `Unsupported import statement: ${statement.moduleSpecifier.getText(
            sourceFile
          )}`
        );
      }
      const importPath = statement.moduleSpecifier.text;
      if (!importPath.startsWith(".")) {
        // This is not a relative import, we'll ignore it.
        continue;
      }
      await parseFileRecursively(
        api,
        visitedPaths,
        path.join(sourcePath, "..", importPath)
      );
    } else if (ts.isClassDeclaration(statement)) {
      const apiDecorator = extractSingleDecorator(sourceFile, statement, "api");
      if (apiDecorator) {
        parseApiClass(sourceFile, statement, api);
      }
    } else if (ts.isTypeAliasDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractType(sourceFile, statement.type);
    } else if (ts.isInterfaceDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractObjectType(sourceFile, statement);
    }
  }
}
