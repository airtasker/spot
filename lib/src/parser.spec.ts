import * as fs from "fs-extra";
import * as path from "path";
import { parsePath } from "./parsing/file-parser";

const EXAMPLES_DIR = path.join(__dirname, "..", "..", "examples", "src");

describe("Parser", () => {
  describe("produces the expected output for example API definitions", async () => {
    for (const testCaseName of fs.readdirSync(EXAMPLES_DIR)) {
      if (!fs.lstatSync(path.join(EXAMPLES_DIR, testCaseName)).isDirectory()) {
        continue;
      }
      test(testCaseName, async () => {
        let parsedApi = await parsePath(
          path.join(EXAMPLES_DIR, testCaseName, "api.ts")
        );
        expect(parsedApi).toMatchSnapshot();
      });
    }
  });
});
