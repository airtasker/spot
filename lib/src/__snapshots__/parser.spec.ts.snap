// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser produces the expected output for example API definitions multi-file 1`] = `
Object {
  "endpoints": Object {
    "createUser": Object {
      "genericErrorType": Object {
        "kind": "void",
      },
      "headers": Object {
        "authToken": Object {
          "headerFieldName": "Authorization",
          "type": Object {
            "kind": "optional",
            "optional": Object {
              "kind": "string",
            },
          },
        },
      },
      "method": "POST",
      "path": Array [
        Object {
          "content": "/users/create",
          "kind": "static",
        },
      ],
      "queryParams": Array [],
      "requestContentType": "application/json",
      "requestType": Object {
        "kind": "type-reference",
        "typeName": "CreateUserRequest",
      },
      "responseType": Object {
        "kind": "type-reference",
        "typeName": "CreateUserResponse",
      },
      "specificErrorTypes": Object {},
    },
    "deleteUser": Object {
      "genericErrorType": Object {
        "kind": "object",
        "properties": Object {
          "message": Object {
            "kind": "string",
          },
        },
      },
      "headers": Object {
        "authToken": Object {
          "headerFieldName": "Authorization",
          "type": Object {
            "kind": "string",
          },
        },
      },
      "method": "DELETE",
      "path": Array [
        Object {
          "content": "/users/",
          "kind": "static",
        },
        Object {
          "kind": "dynamic",
          "name": "userId",
          "type": Object {
            "kind": "string",
          },
        },
        Object {
          "content": "-confirmed",
          "kind": "static",
        },
      ],
      "queryParams": Array [],
      "requestContentType": "application/json",
      "requestType": Object {
        "kind": "void",
      },
      "responseType": Object {
        "kind": "null",
      },
      "specificErrorTypes": Object {
        "forbidden": Object {
          "statusCode": 403,
          "type": Object {
            "kind": "object",
            "properties": Object {
              "message": Object {
                "kind": "string",
              },
              "signedInAs": Object {
                "kind": "string",
              },
            },
          },
        },
      },
    },
    "findUsers": Object {
      "genericErrorType": Object {
        "kind": "void",
      },
      "headers": Object {},
      "method": "GET",
      "path": Array [
        Object {
          "content": "/users",
          "kind": "static",
        },
      ],
      "queryParams": Array [
        Object {
          "name": "limit",
          "type": Object {
            "kind": "number",
          },
        },
        Object {
          "name": "search_term",
          "type": Object {
            "kind": "optional",
            "optional": Object {
              "kind": "string",
            },
          },
        },
      ],
      "requestContentType": "application/json",
      "requestType": Object {
        "kind": "void",
      },
      "responseType": Object {
        "elements": Object {
          "kind": "object",
          "properties": Object {
            "age": Object {
              "kind": "optional",
              "optional": Object {
                "kind": "number",
              },
            },
            "name": Object {
              "kind": "string",
            },
          },
        },
        "kind": "array",
      },
      "specificErrorTypes": Object {},
    },
    "getUser": Object {
      "genericErrorType": Object {
        "kind": "void",
      },
      "headers": Object {},
      "method": "GET",
      "path": Array [
        Object {
          "content": "/users/",
          "kind": "static",
        },
        Object {
          "kind": "dynamic",
          "name": "userId",
          "type": Object {
            "kind": "string",
          },
        },
      ],
      "queryParams": Array [],
      "requestContentType": "application/json",
      "requestType": Object {
        "kind": "void",
      },
      "responseType": Object {
        "kind": "object",
        "properties": Object {
          "age": Object {
            "kind": "optional",
            "optional": Object {
              "kind": "number",
            },
          },
          "name": Object {
            "kind": "string",
          },
        },
      },
      "specificErrorTypes": Object {},
    },
  },
  "types": Object {
    "CreateUserRequest": Object {
      "kind": "object",
      "properties": Object {
        "name": Object {
          "kind": "string",
        },
      },
    },
    "CreateUserResponse": Object {
      "kind": "object",
      "properties": Object {
        "success": Object {
          "kind": "boolean",
        },
      },
    },
    "GetUserResponse": Object {
      "kind": "object",
      "properties": Object {
        "name": Object {
          "kind": "string",
        },
      },
    },
  },
}
`;

exports[`Parser produces the expected output for example API definitions single-file 1`] = `
Object {
  "endpoints": Object {
    "createUser": Object {
      "genericErrorType": Object {
        "kind": "void",
      },
      "headers": Object {
        "authToken": Object {
          "headerFieldName": "Authorization",
          "type": Object {
            "kind": "optional",
            "optional": Object {
              "kind": "string",
            },
          },
        },
      },
      "method": "POST",
      "path": Array [
        Object {
          "content": "/users/create",
          "kind": "static",
        },
      ],
      "queryParams": Array [],
      "requestContentType": "application/json",
      "requestType": Object {
        "kind": "type-reference",
        "typeName": "CreateUserRequest",
      },
      "responseType": Object {
        "kind": "type-reference",
        "typeName": "CreateUserResponse",
      },
      "specificErrorTypes": Object {},
    },
    "deleteUser": Object {
      "genericErrorType": Object {
        "kind": "object",
        "properties": Object {
          "message": Object {
            "kind": "string",
          },
        },
      },
      "headers": Object {
        "authToken": Object {
          "headerFieldName": "Authorization",
          "type": Object {
            "kind": "string",
          },
        },
      },
      "method": "DELETE",
      "path": Array [
        Object {
          "content": "/users/",
          "kind": "static",
        },
        Object {
          "kind": "dynamic",
          "name": "userId",
          "type": Object {
            "kind": "string",
          },
        },
        Object {
          "content": "-confirmed",
          "kind": "static",
        },
      ],
      "queryParams": Array [],
      "requestContentType": "application/json",
      "requestType": Object {
        "kind": "void",
      },
      "responseType": Object {
        "kind": "null",
      },
      "specificErrorTypes": Object {
        "forbidden": Object {
          "statusCode": 403,
          "type": Object {
            "kind": "object",
            "properties": Object {
              "message": Object {
                "kind": "string",
              },
              "signedInAs": Object {
                "kind": "string",
              },
            },
          },
        },
      },
    },
    "findUsers": Object {
      "genericErrorType": Object {
        "kind": "void",
      },
      "headers": Object {},
      "method": "GET",
      "path": Array [
        Object {
          "content": "/users",
          "kind": "static",
        },
      ],
      "queryParams": Array [
        Object {
          "name": "limit",
          "type": Object {
            "kind": "number",
          },
        },
        Object {
          "name": "search_term",
          "type": Object {
            "kind": "optional",
            "optional": Object {
              "kind": "string",
            },
          },
        },
      ],
      "requestContentType": "application/json",
      "requestType": Object {
        "kind": "void",
      },
      "responseType": Object {
        "elements": Object {
          "kind": "object",
          "properties": Object {
            "age": Object {
              "kind": "optional",
              "optional": Object {
                "kind": "number",
              },
            },
            "name": Object {
              "kind": "string",
            },
          },
        },
        "kind": "array",
      },
      "specificErrorTypes": Object {},
    },
    "getUser": Object {
      "genericErrorType": Object {
        "kind": "void",
      },
      "headers": Object {},
      "method": "GET",
      "path": Array [
        Object {
          "content": "/users/",
          "kind": "static",
        },
        Object {
          "kind": "dynamic",
          "name": "userId",
          "type": Object {
            "kind": "string",
          },
        },
      ],
      "queryParams": Array [],
      "requestContentType": "application/json",
      "requestType": Object {
        "kind": "void",
      },
      "responseType": Object {
        "kind": "object",
        "properties": Object {
          "age": Object {
            "kind": "optional",
            "optional": Object {
              "kind": "number",
            },
          },
          "name": Object {
            "kind": "string",
          },
        },
      },
      "specificErrorTypes": Object {},
    },
  },
  "types": Object {
    "CreateUserRequest": Object {
      "kind": "object",
      "properties": Object {
        "name": Object {
          "kind": "string",
        },
      },
    },
    "CreateUserResponse": Object {
      "kind": "object",
      "properties": Object {
        "success": Object {
          "kind": "boolean",
        },
      },
    },
    "GetUserResponse": Object {
      "kind": "object",
      "properties": Object {
        "name": Object {
          "kind": "string",
        },
      },
    },
  },
}
`;
